File: dashboard/services_control.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// services_control.go -- the dashboard's only path to hp-services-adapter
// (#197). Mirrors workbench_model_status.go's transport exactly: a fixed
// local AF_UNIX socket, never a browser-supplied address, with the response
// schema validated field-by-field before anything here is trusted. Unlike
// the read-only model-status adapter, this one accepts actions (start/stop/
// restart) -- the adapter itself holds the real allowlist and is the only
// thing with docker.sock; this client just carries the request and refuses
// to forward anything the adapter's own schema doesn't recognize.

type serviceStatus struct {
	Name         string `json:"name"`
	State        string `json:"state"`
	Health       string `json:"health,omitempty"`
	ExitCode     *int   `json:"exit_code"`
	StartedAt    string `json:"started_at,omitempty"`
	RestartCount *int   `json:"restart_count"`
}

var validServiceStates = map[string]bool{
	"running": true, "exited": true, "restarting": true, "paused": true,
	"created": true, "removing": true, "dead": true, "not_found": true, "unknown": true,
}

func validServiceStatus(s serviceStatus) bool {
	return strings.TrimSpace(s.Name) != "" && validServiceStates[s.State]
}

func servicesAdapterSocket() (string, bool) {
	socket := filepath.Clean(strings.TrimSpace(os.Getenv("SERVICES_ADAPTER_SOCKET")))
	if socket == "." || !filepath.IsAbs(socket) || strings.Contains(socket, "..") {
		return "", false
	}
	return socket, true
}

func servicesAdapterClient(timeout time.Duration) (*http.Client, string, bool) {
	socket, ok := servicesAdapterSocket()
	if !ok {
		return nil, "", false
	}
	transport := &http.Transport{DialContext: func(ctx context.Context, _, _ string) (net.Conn, error) {
		return (&net.Dialer{Timeout: time.Second}).DialContext(ctx, "unix", socket)
	}}
	client := &http.Client{Transport: transport, Timeout: timeout, CheckRedirect: func(_ *http.Request, _ []*http.Request) error { return http.ErrUseLastResponse }}
	return client, socket, true
}

// loadServicesStatus fetches the current allowlisted-container inventory
// from hp-services-adapter. The bool return distinguishes "the adapter isn't
// configured or reachable" from "it returned zero services" -- callers must
// not render an empty pane as if it were an authoritative empty state.
func loadServicesStatus() ([]serviceStatus, bool, string) {
	client, _, ok := servicesAdapterClient(8 * time.Second)
	if !ok {
		return nil, false, "services adapter is not configured"
	}
	response, err := client.Get("http://services-adapter/v1/services")
	if err != nil {
		return nil, false, "services adapter is unavailable"
	}
	defer response.Body.Close()
	if response.StatusCode != http.StatusOK || !strings.HasPrefix(strings.ToLower(response.Header.Get("Content-Type")), "application/json") {
		return nil, false, "services adapter returned an invalid response"
	}
	var payload struct {
		Services []serviceStatus `json:"services"`
	}
	decoder := json.NewDecoder(io.LimitReader(response.Body, 256<<10))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&payload); err != nil {
		return nil, false, "services adapter returned an invalid schema"
	}
	for _, service := range payload.Services {
		if !validServiceStatus(service) {
			return nil, false, "services adapter returned an invalid schema"
		}
	}
	return payload.Services, true, ""
}

var validServiceActions = map[string]bool{"start": true, "stop": true, "restart": true}

// performServiceAction asks hp-services-adapter to start/stop/restart one
// container. The adapter itself enforces the real allowlist and returns 403
// for anything not on it -- this only pre-validates the action verb so an
// obviously-bad request never leaves the dashboard.
func performServiceAction(name, action string) (bool, int, string) {
	if strings.TrimSpace(name) == "" || !validServiceActions[action] {
		return false, http.StatusBadRequest, "unsupported action"
	}
	client, _, ok := servicesAdapterClient(15 * time.Second)
	if !ok {
		return false, http.StatusServiceUnavailable, "services adapter is not configured"
	}
	request, err := http.NewRequest(http.MethodPost, "http://services-adapter/v1/services/"+url.PathEscape(name)+"/"+action, nil)
	if err != nil {
		return false, http.StatusInternalServerError, "could not build request"
	}
	response, err := client.Do(request)
	if err != nil {
		return false, http.StatusServiceUnavailable, "services adapter is unavailable"
	}
	defer response.Body.Close()
	var payload struct {
		OK     bool   `json:"ok"`
		Action string `json:"action"`
		Name   string `json:"name"`
		Error  string `json:"error"`
	}
	decoder := json.NewDecoder(io.LimitReader(response.Body, 16<<10))
	decoder.DisallowUnknownFields()
	_ = decoder.Decode(&payload)
	if response.StatusCode != http.StatusOK || !payload.OK {
		reason := payload.Error
		if reason == "" {
			reason = fmt.Sprintf("services adapter returned %d", response.StatusCode)
		}
		return false, response.StatusCode, reason
	}
	return true, http.StatusOK, ""
}

const maxServiceLogLines = 1000

// fetchServiceLogs tails one container's log through hp-services-adapter.
// lines is clamped the same way the adapter itself clamps it -- belt and
// suspenders, since the adapter is the actual enforcement point.
func fetchServiceLogs(name string, lines int) (string, int, string) {
	if strings.TrimSpace(name) == "" {
		return "", http.StatusBadRequest, "missing service name"
	}
	if lines <= 0 {
		lines = 200
	}
	if lines > maxServiceLogLines {
		lines = maxServiceLogLines
	}
	client, _, ok := servicesAdapterClient(15 * time.Second)
	if !ok {
		return "", http.StatusServiceUnavailable, "services adapter is not configured"
	}
	response, err := client.Get(fmt.Sprintf("http://services-adapter/v1/services/%s/logs?lines=%d", url.PathEscape(name), lines))
	if err != nil {
		return "", http.StatusServiceUnavailable, "services adapter is unavailable"
	}
	defer response.Body.Close()
	if response.StatusCode != http.StatusOK {
		return "", response.StatusCode, fmt.Sprintf("services adapter returned %d", response.StatusCode)
	}
	var payload struct {
		Name  string `json:"name"`
		Lines int    `json:"lines"`
		Log   string `json:"log"`
	}
	decoder := json.NewDecoder(io.LimitReader(response.Body, 4<<20))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&payload); err != nil {
		return "", http.StatusBadGateway, "services adapter returned an invalid schema"
	}
	return payload.Log, http.StatusOK, ""
}

File: arcane/home/honeypot-dashboard/backend-service/src/services_control.rs
//! Services pane API (#197), ported from settings_services_api.go /
//! services_control.go: the only path to hp-services-adapter. The real
//! allowlist and the real docker.sock access live only in the adapter
//! (services-adapter/services-adapter.py) — this module only ever forwards
//! a name and an action verb over the fixed AF_UNIX socket
//! (SERVICES_ADAPTER_SOCKET), and never gains any other path to Docker.
//!
//! The Go tier talks to the socket through Go's stdlib http.Client with a
//! custom unix dialer. There is no equivalent one-liner in this crate's
//! dependency set (reqwest has no unix-socket transport), so this is a
//! small hand-rolled HTTP/1.1 client over tokio::net::UnixStream rather
//! than pulling in a new dependency for one internal, trusted, small-JSON
//! call — every response is read to EOF after requesting `Connection:
//! close`, bounded by max_body, which sidesteps needing a real
//! Content-Length/chunked-transfer parser.

use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    Json,
};
use serde::Deserialize;
use serde_json::{json, Value};
use std::time::Duration;
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::UnixStream;

use crate::audit::AuditEvent;
use crate::AppState;

const MAX_SERVICE_LOG_LINES: u32 = 1000;
const VALID_STATES: &[&str] =
    &["running", "exited", "restarting", "paused", "created", "removing", "dead", "not_found", "unknown"];
const VALID_ACTIONS: &[&str] = &["start", "stop", "restart"];

fn socket_path() -> Option<String> {
    std::env::var("SERVICES_ADAPTER_SOCKET").ok().filter(|s| !s.trim().is_empty())
}

async fn unix_request(
    socket: &str,
    method: &str,
    path: &str,
    body: Option<&[u8]>,
    timeout: Duration,
    max_body: usize,
) -> anyhow::Result<(u16, Vec<u8>)> {
    let mut stream = tokio::time::timeout(Duration::from_secs(1), UnixStream::connect(socket))
        .await
        .map_err(|_| anyhow::anyhow!("services adapter connect timeout"))??;

    let mut request = format!(
        "{method} {path} HTTP/1.1\r\nHost: services-adapter\r\nConnection: close\r\nContent-Length: {}\r\n",
        body.map(|b| b.len()).unwrap_or(0)
    );
    if body.is_some() {
        request.push_str("Content-Type: application/json\r\n");
    }
    request.push_str("\r\n");
    let mut out = request.into_bytes();
    if let Some(body) = body {
        out.extend_from_slice(body);
    }
    tokio::time::timeout(timeout, stream.write_all(&out)).await??;
    stream.shutdown().await.ok();

    let mut raw = Vec::new();
    tokio::time::timeout(timeout, stream.read_to_end(&mut raw)).await??;

    let header_end = raw
        .windows(4)
        .position(|w| w == b"\r\n\r\n")
        .ok_or_else(|| anyhow::anyhow!("services adapter returned a malformed response"))?;
    let header_text = std::str::from_utf8(&raw[..header_end])?;
    let status_line = header_text.lines().next().unwrap_or("");
    let status: u16 = status_line
        .split_whitespace()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .ok_or_else(|| anyhow::anyhow!("services adapter returned a malformed status line"))?;

    let mut response_body = raw[header_end + 4..].to_vec();
    if response_body.len() > max_body {
        response_body.truncate(max_body);
    }
    Ok((status, response_body))
}

/// Fetches the current allowlisted-container inventory. Ok(None) means the
/// adapter isn't configured/reachable/valid — distinct from "zero
/// services" — so callers never render an empty pane as authoritative.
async fn load_services_status() -> Result<Vec<Value>, String> {
    let Some(socket) = socket_path() else { return Err("services adapter is not configured".into()) };
    let (status, body) = unix_request(&socket, "GET", "/v1/services", None, Duration::from_secs(8), 256 << 10)
        .await
        .map_err(|_| "services adapter is unavailable".to_string())?;
    if status != 200 {
        return Err("services adapter returned an invalid response".into());
    }
    let payload: Value =
        serde_json::from_slice(&body).map_err(|_| "services adapter returned an invalid schema".to_string())?;
    let services = payload["services"].as_array().cloned().unwrap_or_default();
    let all_valid = services.iter().all(|service| {
        let name_ok = service["name"].as_str().is_some_and(|n| !n.trim().is_empty());
        let state_ok = service["state"].as_str().is_some_and(|s| VALID_STATES.contains(&s));
        name_ok && state_ok
    });
    if !all_valid {
        return Err("services adapter returned an invalid schema".into());
    }
    Ok(services)
}

async fn perform_service_action(name: &str, action: &str) -> (bool, u16, String) {
    if name.trim().is_empty() || !VALID_ACTIONS.contains(&action) {
        return (false, 400, "unsupported action".into());
    }
    let Some(socket) = socket_path() else {
        return (false, 503, "services adapter is not configured".into());
    };
    let path = format!("/v1/services/{}/{}", urlencode(name), action);
    let Ok((status, body)) = unix_request(&socket, "POST", &path, Some(b"{}"), Duration::from_secs(15), 16 << 10).await
    else {
        return (false, 503, "services adapter is unavailable".into());
    };
    let payload: Value = serde_json::from_slice(&body).unwrap_or(Value::Null);
    let ok = status == 200 && payload["ok"].as_bool().unwrap_or(false);
    if ok {
        return (true, 200, String::new());
    }
    let reason = payload["error"].as_str().filter(|s| !s.is_empty()).map(String::from);
    (false, status, reason.unwrap_or_else(|| format!("services adapter returned {status}")))
}

async fn fetch_service_logs(name: &str, lines: u32) -> Result<(String, u32), (u16, String)> {
    if name.trim().is_empty() {
        return Err((400, "missing service name".into()));
    }
    let lines = if lines == 0 { 200 } else { lines.min(MAX_SERVICE_LOG_LINES) };
    let Some(socket) = socket_path() else {
        return Err((503, "services adapter is not configured".into()));
    };
    let path = format!("/v1/services/{}/logs?lines={}", urlencode(name), lines);
    let (status, body) = unix_request(&socket, "GET", &path, None, Duration::from_secs(15), 4 << 20)
        .await
        .map_err(|_| (503u16, "services adapter is unavailable".to_string()))?;
    if status != 200 {
        return Err((status, format!("services adapter returned {status}")));
    }
    let payload: Value =
        serde_json::from_slice(&body).map_err(|_| (502u16, "services adapter returned an invalid schema".to_string()))?;
    let log = payload["log"].as_str().unwrap_or_default().to_string();
    Ok((log, lines))
}

pub fn urlencode(value: &str) -> String {
    let mut out = String::with_capacity(value.len());
    for byte in value.bytes() {
        match byte {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => out.push(byte as char),
            _ => out.push_str(&format!("%{byte:02X}")),
        }
    }
    out
}

pub async fn list(State(_state): State<AppState>) -> impl axum::response::IntoResponse {
    match load_services_status().await {
        Ok(services) => (StatusCode::OK, Json(json!({"available": true, "services": services}))),
        Err(reason) => (
            StatusCode::SERVICE_UNAVAILABLE,
            Json(json!({"available": false, "reason": reason, "services": []})),
        ),
    }
}

#[derive(Deserialize)]
pub struct LogsQuery {
    lines: Option<u32>,
}

pub async fn logs(
    State(_state): State<AppState>,
    Path(name): Path<String>,
    Query(query): Query<LogsQuery>,
) -> impl axum::response::IntoResponse {
    match fetch_service_logs(&name, query.lines.unwrap_or(200)).await {
        Ok((log, lines)) => (StatusCode::OK, Json(json!({"name": name, "lines": lines, "log": log}))),
        Err((status, reason)) => (
            StatusCode::from_u16(status).unwrap_or(StatusCode::BAD_GATEWAY),
            Json(json!({"error": reason})),
        ),
    }
}

#[derive(Deserialize)]
pub struct ActionQuery {
    #[serde(default)]
    actor_subject: String,
    #[serde(default)]
    actor_username: String,
}

pub async fn action(
    State(state): State<AppState>,
    Path((name, action)): Path<(String, String)>,
    Query(actor): Query<ActionQuery>,
) -> impl axum::response::IntoResponse {
    let (ok, status, reason) = perform_service_action(&name, &action).await;
    state.audit.log(AuditEvent {
        actor_subject: actor.actor_subject,
        actor_username: actor.actor_username,
        action: format!("services.{action}"),
        fields: vec![name.clone()],
        result: if ok { "success" } else { "error" }.into(),
        ..Default::default()
    });
    let body = if ok { json!({"ok": true, "name": name, "action": action}) } else { json!({"ok": false, "error": reason}) };
    (StatusCode::from_u16(status).unwrap_or(StatusCode::BAD_GATEWAY), Json(body))
}
